### ISSUE TITLE
[Java][Spring] Optional Parameters not in Generated Swagger and SwaggerDocumentationConfig bean not being used

### ISSUE THREAD

**Issue by slarti-b** [swagger-api/swagger-codegen#7601]
##### Description

Generating a Java Spring server stub.  The swagger file I generate from includes a path with a post with optional query parameters.  The annotations generated look OK to me, but the swagger file generated from the code (endpoint specified as `springfox.documentation.swagger.v2.path` in the application.properties file) does not include the optional Query parameters

E.g. generating from a swagger including:
```yaml
parameters:
  fooQuery:
    name: fooQuery
    in: query
    type: string
    required: false
paths:
  /foo/bar:
    post:
      operationId: postBar
      tags:
        - foo
      parameters:
        - $ref: '#/parameters/fooQuery'        
```

The generated code looks OK in terms of annotations:
```java
@Api(value = "foo", description = "the foo API")
public interface FooApi {

    FooApiDelegate getDelegate();

    @ApiOperation(value = "", nickname = "postBar", notes = "", tags={ "foo", })
    @ApiResponses(value = {  })
    @RequestMapping(value = "/foo/bar",
        produces = { "application/json" }, 
        consumes = { "application/json" },
        method = RequestMethod.POST)
    default CompletableFuture<ResponseEntity<Void>> postBar(@ApiParam(value = "") @Valid @RequestParam(value = "fooQuery", required = false) Optional<String> fooQuery) {
        return getDelegate().postBar(fooQuery);
    }

}
```

But the swagger definition generated from the server code misses the `fooQuery` parameter:
```json
	"paths": {
		"/foo/bar": {
			"post": {
				"tags": ["foo"],
				"summary": "postBar",
				"operationId": "postBar",
				"consumes": ["application/json"],
				"produces": ["application/json"],
				"responses": {
					"200": {
						"description": "OK",
						"schema": {
							"$ref": "#/definitions/CompletableFuture«ResponseEntity«Void»»"
						}
					},
					"201": {
						"description": "Created"
					},
					"401": {
						"description": "Unauthorized"
					},
					"403": {
						"description": "Forbidden"
					},
					"404": {
						"description": "Not Found"
					}
				}
			}
		}
	}
```

##### Swagger-codegen version

2.3.0

##### Swagger declaration file content or url

```yaml
swagger: '2.0'
host: api.example.com
basePath: /
schemes:
  - https
consumes:
  - application/json
produces:
  - application/json
tags:
  - name: foo
parameters:
  fooQuery:
    name: fooQuery
    in: query
    type: string
    required: false
paths:
  /foo/bar:
    post:
      operationId: postBar
      tags:
        - foo
      parameters:
        - $ref: '#/parameters/fooQuery'        
```

##### Command line used for generation

I'm generating with the swagger maven plugin.  Relevant config options are:
```xml
<serializableModel>true</serializableModel>
<dateLibrary>java8</dateLibrary>
<java8>true</java8>
<async>true</async>
<library>spring-boot</library>		
<delegatePattern>true</delegatePattern>
<useBeanValidation>true</useBeanValidation>
<useOptional>true</useOptional>	
```

##### Steps to reproduce

Gene

- rate the server stub from the yaml
- Run it
- Go to the endpoint specified as `springfox.documentation.swagger.v2.path` in the application.properties file

##### Suggest a fix/enhancement

I thought it might be related to this springfox issue: https://github.com/springfox/springfox/issues/1848  I tried adding `.genericModelSubstitutes(Optional.class)` to the `customImplementation` function in the generated `SwaggerDocumentationConfig` giving:
```java
package io.swagger.configuration;

import java.util.Optional;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import springfox.documentation.builders.ApiInfoBuilder;
import springfox.documentation.builders.RequestHandlerSelectors;
import springfox.documentation.service.ApiInfo;
import springfox.documentation.service.Contact;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spring.web.plugins.Docket;


@Configuration
public class SwaggerDocumentationConfig {

    ApiInfo apiInfo() {
        return new ApiInfoBuilder()
            .title("")
            .description("")
            .license("")
            .licenseUrl("http://unlicense.org")
            .termsOfServiceUrl("")
            .version("")
            .contact(new Contact("","", ""))
            .build();
    }

    @Bean
    public Docket customImplementation(){
        return new Docket(DocumentationType.SWAGGER_2)
                .select()
                    .apis(RequestHandlerSelectors.basePackage("com.carus.api.push.messages"))
                    .build()
                    .genericModelSubstitutes(Optional.class)
                .directModelSubstitute(java.time.LocalDate.class, java.sql.Date.class)
                .directModelSubstitute(java.time.OffsetDateTime.class, java.util.Date.class)
                .apiInfo(apiInfo());
    }

}
```

But it didn't help.  That answer seems to imply that springfox copes with optional query params, so I'm guessing it's something missing in a template somewhere in the generation, but I'm not sure.

**Comment by slarti-b** [comment]
It's definitely connected to the use of Optional.  Setting `useOptional` to false _or_ making the parameter required _or_ manually changing the generated file to use `String` instead of `Optional<String>` - any of these "fix" the problem and the parameter is used in the generated swagger.  What I don't understand is why adding the genericModelSubstitute doesn't fix it.

**Comment by slarti-b** [comment]
I've done some more digging.  It seems the SwaggerDocumentationConfig file isn't being used at all.  At least, the `customImplementation` bean.  Which is why my fix to add `.genericModelSubstitutes(Optional.class)` to that doesn't help.

I changed the generated SwaggerDocumentationConfig .java to
```java
@Configuration
public class SwaggerDocumentationConfig {

    ApiInfo apiInfo() {
        return new ApiInfoBuilder()
            .title("")
            .description("")
            .license("")
            .licenseUrl("http://unlicense.org")
            .termsOfServiceUrl("")
            .version("")
            .contact(new Contact("","", ""))
            .build();
    }

    @Bean
    public Docket customImplementation(){
    	throw new RuntimeException("qwerty");
//        return new Docket(DocumentationType.SWAGGER_2)
//                .select()
//                    .apis(RequestHandlerSelectors.basePackage("com.carus.api.push.messages"))
//                    .build()
//                    .useDefaultResponseMessages(false)
////                .directModelSubstitute(java.time.LocalDate.class, java.sql.Date.class)
////                .directModelSubstitute(java.time.OffsetDateTime.class, java.util.Date.class)
//				.genericModelSubstitutes(Optional.class)
//                .apiInfo(apiInfo());
    }

}
```

And the exception is never thrown!  Adding the same function/bean to the Swagger2Springboot class (which is how the [springfox docs](https://springfox.github.io/springfox/docs/current/) do it) fixes the problem (with the `.genericModelSubstitutes(Optional.class)` - this is still required)

**Comment by slarti-b** [comment]
I have now found the problem.  I had specified `apiPackage` and `basePackage` but not `configPackage` in my config when I ran codegen.  In this case, the config still ends up under `io.swagger.configuration`, which is *not* referenced as a base package to the `@componentScans` in Swagger2SpringBoot.  So it's not scanned, and the documentation config is ignored.  Will do a PR which:
* Adds configPackage to the scans
* Adds the `genericModelSubstitutes` to the documentation config
